L1Trigger/L1THGCal/src/veryfrontend/HGCalVFECompressionImpl.cc

```diff
@@ -1,5 +1,89 @@
 #include "L1Trigger/L1THGCal/interface/veryfrontend/HGCalVFECompressionImpl.h"
 
 HGCalVFECompressionImpl::
-HGCalVFECompressionImpl(const edm::ParameterSet& conf)
-{}
+HGCalVFECompressionImpl(const edm::ParameterSet& conf):
+    exponentBits_(conf.getParameter<uint32_t>("exponentBits")),
+    mantissaBits_(conf.getParameter<uint32_t>("mantissaBits")),
+    rounding_(conf.getParameter<bool>("rounding"))
+{
+  if (((1 << exponentBits_) + mantissaBits_ - 1) >= 32) {
+    throw cms::Exception("CodespaceCannotFit")
+      << "The code space cannot fit into the unsigned 32-bit space.\n";
+  }
+  saturationCode_ = (1 << (exponentBits_ + mantissaBits_)) - 1;
+  saturationValue_ = (1 << ((1 << exponentBits_) + mantissaBits_ - 1)) - 1;
+  // TODO: for non-saturating need to get maximum code point as well?
+  // i.e. 8-bit compressed code that would hit 0xffffffff
+}
+
+void
+HGCalVFECompressionImpl::
+compressSingle(const uint32_t value,
+               uint32_t & compressedCode,
+               uint32_t & compressedValue)
+{
+  // check for saturation
+  if (value > saturationValue_) {
+    compressedCode = saturationCode_;
+    compressedValue =
```

Comment:
Shouldn't it be set equal to the `saturationValue_`?

Reply:
They are different.

For `saturationValue_` I followed the Python code given by you on the 1st email about this task, in which it is on the uncompressed value space at 1 value just before the codepoint in which it would overflow. I.e. for 4 exponent 4 mantissa, this `saturationValue_` would be 0x7ffff. Or if in codespace there is 0x100 then that would correspond to 0x80000.

While what I am setting for `compressedValue` is the value in uncompressed space corresponding to the largest representable in code space, i.e. in 4e4m case that would be 0x7c000.

Maybe this 0x7c000 should be set as `saturationValue_` instead?